nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+    // running snapshot
+    template_builder.append(
+        "created_volume",
+        "CreateVolumeRecord",
+        ActionFunc::new_action(
+            ssc_create_volume_record,
+            ssc_create_volume_record_undo,
+        ),
+    );
+
+    template_builder.append(
+        "finalized_snapshot",
+        "FinalizeSnapshotRecord",
+        new_action_noop_undo(ssc_finalize_snapshot_record),
+    );
+
+    template_builder.build()
+}
+
+async fn ssc_create_snapshot_record(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<db::model::Snapshot, ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    // We admittedly reference the volume before it has been allocated,
+    // but this should be acceptable because the snapshot remains in a "Creating"
+    // state until the saga has completed.
+    let volume_id = sagactx.lookup::<Uuid>("volume_id")?;
+
+    warn!(log, "grabbing disk by name {}", params.create_params.disk);
+
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    warn!(log, "creating snapshot {} from disk {}", snapshot_id, disk.id());
+
+    let snapshot = db::model::Snapshot {
+        identity: db::model::SnapshotIdentity::new(
+            snapshot_id,
+            params.create_params.identity.clone(),
+        ),
+
+        project_id: params.project_id,
+        disk_id: disk.id(),
+        volume_id,
+
+        gen: db::model::Generation::new(),
+        state: db::model::SnapshotState::Creating,
+        block_size: disk.block_size,
+        size: disk.size,
+    };
+
+    let (authz_silo, ..) = LookupPath::new(&opctx, &osagactx.datastore())
+        .silo_id(params.silo_id)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let snapshot_created = osagactx
+        .datastore()
+        .project_create_snapshot(&opctx, &authz_silo, snapshot)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    warn!(log, "created snapshot {} ok", snapshot_id);
+
+    Ok(snapshot_created)
+}
+
+async fn ssc_create_snapshot_record_undo(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+    warn!(log, "deleting snapshot {}", snapshot_id);
+
+    let (.., authz_snapshot, db_snapshot) =
```

Comment:
Actions and undo actions *both* need to be idempotent for sagas to operate correctly.

If we imagine re-running this node twice, would the second invocation fail and return a "not found" error? If so, it should probably eat that error and return `Ok(())` instead.

Reply:
addressed this in a8bcf1ec2aa2d3981329a5d07b87c0b5e7c32959 and e149e67e595c380c2bab03789b4c34ac79276f9b